app/drivers/kscan/kscan_gpio_matrix.c

```diff
@@ -51,257 +158,300 @@ static int kscan_gpio_config_interrupts(const struct device **devices,
-            CONTAINER_OF(cb, struct kscan_gpio_irq_callback_##n, callback);                        \
-        kscan_gpio_disable_interrupts_##n(data->dev);                                              \
-        COND_CODE_0(DT_INST_PROP(n, debounce_period), ({ k_work_submit(data->work); }), ({         \
-                        k_delayed_work_cancel(data->work);                                         \
-                        k_delayed_work_submit(data->work,                                          \
-                                              K_MSEC(DT_INST_PROP(n, debounce_period)));           \
-                    }))                                                                            \
-    })                                                                                             \
+    // While interrupts are enabled, set all outputs active so a pressed key
+    // will trigger an interrupt.
+    return kscan_matrix_set_all_outputs(dev, 1);
+}
+#endif
+
+#if USE_INTERRUPTS
+static int kscan_matrix_interrupt_disable(const struct device *dev) {
+    int err = kscan_matrix_interrupt_configure(dev, GPIO_INT_DISABLE);
+    if (err) {
+        return err;
+    }
+
+    // While interrupts are disabled, set all outputs inactive so
+    // kscan_matrix_read() can scan them one by one.
+    return kscan_matrix_set_all_outputs(dev, 0);
+}
+#endif
+
+#if USE_INTERRUPTS
+static void kscan_matrix_irq_callback_handler(const struct device *port, struct gpio_callback *cb,
+                                              const gpio_port_pins_t pin) {
+    struct kscan_matrix_irq_callback *data =
+        CONTAINER_OF(cb, struct kscan_matrix_irq_callback, callback);
+    const struct kscan_matrix_config *config = data->dev->config;
+
+    // Disable our interrupts temporarily to avoid re-entry while we scan.
+    kscan_matrix_interrupt_disable(data->dev);
+
+    // TODO (Zephyr 2.6): use k_work_reschedule()
+    k_delayed_work_cancel(data->work);
+    k_delayed_work_submit(data->work, K_MSEC(config->debounce_period_ms));
+}
+#endif
+
+static int kscan_matrix_read(const struct device *dev) {
+    struct kscan_matrix_data *data = dev->data;
+    const struct kscan_matrix_config *config = dev->config;
+
+    // Scan the matrix.
+    for (int o = 0; o < config->outputs.len; o++) {
+        const struct kscan_gpio_dt_spec *out_gpio = &config->outputs.gpios[o];
+
+        int err = gpio_pin_set(out_gpio->port, out_gpio->pin, 1);
+        if (err) {
+            LOG_ERR("Failed to set output %i active: %i", o, err);
+            return err;
+        }
+
+        for (int i = 0; i < config->inputs.len; i++) {
+            const struct kscan_gpio_dt_spec *in_gpio = &config->inputs.gpios[i];
+
+            const int index = state_index_io(config, i, o);
+            const int pressed = gpio_pin_get(in_gpio->port, in_gpio->pin);
+            data->next_state[index] = pressed;
```

Comment:
Perhaps:

```suggestion
            data->next_state[index] = gpio_pin_get(in_gpio->port, in_gpio->pin);
```

Reply:
Done. (At one point line 222 was a function call like in the old code, but I changed it and missed simplifying it further.)